src/frontend/src/players/vlc/messaging.ts

```diff
@@ -0,0 +1,131 @@
+import type { ChildProcess } from 'child_process'
+import { EOL } from 'os'
+
+enum MessageType {
+  VLC = 'vlc',
+  LOG = 'log',
+  ERROR = 'error',
+  UNKNOWN = 'unknown',
+  ALIVE = 'alive',
+}
+
+enum VLCLogLevel {
+  DEBUG = 'Debug',
+  WARN = 'Warning',
+  ERROR = 'Error',
+  INFO = 'Info',
+}
+
+enum MessageDataType {
+  PLAY = 'play',
+  PAUSE = 'pause',
+  STOP = 'stop',
+  RESUME = 'resume',
+  SET_SUBTITLES = 'subtitles',
+  SET_AUDIO = 'audio',
+  RESIZE = 'resize',
+}
+
+interface IMessageData {
+  type?: MessageDataType | string
+  value: string
+}
+
+interface IMessage {
+  type: MessageType
+  data: IMessageData
+}
+
+interface LogMessage {
+  Timestamp: string
+  Level: VLCLogLevel
+  MessageTemplate: string
+}
+
+class Message implements IMessage {
+  static FromJSON(message: string): Message {
+    try {
+      const deserialized = JSON.parse(message) as IMessage
+      return new Message(deserialized.type || MessageType.LOG, deserialized.data ?? { value: message })
+    } catch (e) {
+      return new Message(MessageType.UNKNOWN, { value: message })
+    }
+  }
+
+  constructor(public readonly type: MessageType, public readonly data: IMessageData) {}
+
+  toJSON(): string {
+    return JSON.stringify({ type: this.type, data: this.data })
+  }
+}
+
+type MessageListener = (message: Message) => void | Promise<void>
+
+class ProcessMessaging {
+  private readonly listeners = new Set<MessageListener>()
+  private readonly bufferedMessages: Message[] = []
+
+  constructor(private childProcess: ChildProcess) {
+    /* eslint-disable @typescript-eslint/no-unsafe-assignment */
+    this.received = this.received.bind(this)
```

Comment:
why do this instead of using the arrow format?

Reply:
Using the arrow format where ? When declaring the functions themselves or in the listeners ?

Reply:
When declaring the functions

Reply:
The semantic is different. The arrow format does not define a function on the class' prototype but creates a property that is a function in the constructor. The class' signature is different. For example you can't use a child's function overridden definition in the parent's constructor when defined with an arrow function.
Apart from the semantic I have nothing else to add